File: example/ios/FlutterLibPDTests/Helpers/SharedTestHelpers.swift
//
//  shared_test_helpers.swift
//  FlutterLibPDTests
//
//  Created by flav on 25/05/2022.
//

import flutter_libpd
import XCTest

extension XCTestCase {
    var timeout: TimeInterval { 0.1 }

    var testBundle: Bundle {
        Bundle(for: type(of: self))
    }
    
    func anyPatchFile(_ patchName: String = "main.pd") -> PatchFile {
        PatchFile(name: patchName, path: getPath(forPatch: patchName, in: testBundle) + "/")
    }

    private func getPath(forPatch name: String, in bundle: Bundle) -> String {
        let splittedName = name.split(separator: ".")
        let path = bundle.path(forResource: "\(splittedName[0])", ofType: "\(splittedName[1])")!
        var url = URL(fileURLWithPath: path)
        url.deleteLastPathComponent()
        return url.path
    }
    
    func anyStringValue() -> String {
        "anyString"
    }
    
    func anyFloatValue() -> Float {
        0.5
    }
    
    func anyReceiver() -> String {
        "anyReceiver"
    }
    
    func anyFloatMessage(_ receiver: String? = nil) -> FloatMessage {
        FloatMessage(value: anyFloatValue(), receiver: receiver ?? anyReceiver())
    }
    
    func anyBangMessage(_ receiver: String? = nil) -> BangMessage {
        BangMessage(receiver: receiver ?? anyReceiver())
    }
    
    func anySymbolMessage(_ receiver: String? = nil) -> SymbolMessage {
        SymbolMessage(value: anyStringValue(), receiver: receiver ?? anyReceiver())
    }
    
    func anyListMessage(_ receiver: String? = nil) -> ListMessage {
        ListMessage(value: [anyFloatValue(), anyStringValue()], receiver: receiver ?? anyReceiver())
    }
    
    func anyBaseMessage(_ receiver: String? = nil) -> ArgumentMessage {
        ArgumentMessage(value: anyStringValue(), arguments: [anyStringValue(), anyFloatValue()], receiver: receiver ?? anyReceiver())
    }
    
    func anyPlaybackConfiguration(inputEnabled: Bool = false) -> PlaybackConfiguration {
        PlaybackConfiguration(sampleRate: 44100,
                              inputNumberOfChannels: 1,
                              outputNumberOfChannels: 1,
                              inputEnabled: inputEnabled)
    }
    
    func makeInvertedExpectation(description: String? = nil) -> XCTestExpectation {
        let expectation = XCTestExpectation(description: description ?? "")
        expectation.isInverted = true
        return expectation
    }
    
    func realReceiver() -> String {
        "pd-gameSlot"
    }
}

File: example/android/app/src/test/java/com/ovaom/flutter_libpd_example/helpers_test/SharedTestHelpers.kt
package com.ovaom.flutter_libpd_example.helpers_test

import com.ovaom.flutter_libpd.pd_controller.pd_audio_manager.PlaybackConfiguration
import com.ovaom.flutter_libpd.pd_controller.pd_messenger.*

fun anyFloatMessage(): FloatMessage {
    return FloatMessage(anyFloatValue(), anyReceiver())
}

fun anyBangMessage(): BangMessage {
    return BangMessage(anyReceiver())
}

fun anySymbolMessage(): SymbolMessage {
    return SymbolMessage(anyStringValue(), anyReceiver())
}

fun anyListMessage(): ListMessage {
    return ListMessage(listOf<Any>(anyFloatValue(), anyStringValue()), anyReceiver())
}

fun anyBaseMessage(): ArgumentMessage {
    return ArgumentMessage(anyStringValue(), listOf<Any>(anyStringValue(), anyFloatValue()), anyReceiver())
}

fun anyFloatValue(): Float {
    return 0.5.toFloat()
}

fun anyReceiver(): String {
    return "anyReceiver"
}

fun anyStringValue(): String {
    return "anyString"
}

fun anyPlaybackConfiguration(): PlaybackConfiguration {
    return PlaybackConfiguration(44100,
            1,
            1)
}